### ISSUE TITLE
What is the purpose of "ratio" in train.py?

### ISSUE THREAD

**Issue by Bill2462** [hangligit/DalleFlamingo#1]
Hello I am trying to implement a modified version of this model but I don't understand what is the purpose of this code located between line 430 and line 441 in `train.py`:

```
 if accelerator.sync_gradients:
      progress_bar.update(1)
      global_step += 1
      if (step+1)%args.gradient_accumulation_steps!=0: 
          ratio=args.gradient_accumulation_steps/((step+1)%args.gradient_accumulation_steps)
          train_loss = train_loss*ratio
          train_loss_lm = train_loss_lm*ratio
          train_loss_sd = train_loss_sd*ratio
      accelerator.log({"train_loss": train_loss, "train_loss_lm": train_loss_lm, "train_loss_sd": train_loss_sd}, step=global_step)
      train_loss = 0.0
      train_loss_lm = 0.0
      train_loss_sd = 0.0

      logs = {"step_loss": loss.detach().item(), "lr_sd": lr_scheduler.get_last_lr()[0], "lr_blip":  optimizer_blip.param_groups[0]['lr']}
      progress_bar.set_postfix(**logs)
```

If I understand correctly those variables:
```
  train_loss += avg_loss.item() / args.gradient_accumulation_steps
  train_loss_lm += loss_lm.item() / args.gradient_accumulation_steps
  train_loss_sd += loss_sd.item() / args.gradient_accumulation_steps
```

will hold average value of respective losses for the given batch. Why this is not just logged and is instead scaled by `ratio`?

**Comment by hangligit** [comment]
Hi this trick is applied only to the last incomplete accumulation cycle when the number of optimization steps is not divisible by the number of accumulation steps. For the complete gradients accumulation cycles, line 434 - line 437 are simply not executed.
